File: js/30-39/problem33.js
const prototypes = require('../prototypes');
const util = require('../util');

var digitCancellingFractions = function() {
    var arr = [];
    var product = 1;

    for (let numerator = 10; numerator < 100; numerator++) {
        for (let denominator = 10; denominator < 100; denominator++) {
            var fraction = numerator / denominator;

            if (fraction >= 1) {
                continue;
            }

            numerator = numerator.toString();
            denominator = denominator.toString();
            var digitCancelFraction;

            // Check for matching non-trivial digits.
            if (numerator[0] === denominator[1] && numerator[0] !== denominator[0]) {
                digitCancelFraction = parseInt(numerator[1]) / parseInt(denominator[0]);
            } else if (numerator[1] === denominator[0] && numerator[1] !== denominator[1]) {
                digitCancelFraction = parseInt(numerator[0]) / parseInt(denominator[1]);
            } else {
                continue;
            }

            if (fraction === digitCancelFraction) {
                arr.push([parseInt(numerator), parseInt(denominator)]);
                product *= fraction;
            }
        }
    }
    console.log(arr)
    return product;
};

util.timeFunc(digitCancellingFractions());

File: python/30-39/problem33.py
'''Not finished yet. I need to write a function to find lowest common terms.'''

def digit_canceling_fractions():
    '''Returns every non-trivial digit canceling fraction.'''
    fractions = []
    for denominator in range(11,100):
        for numerator in range(10, denominator):

            # Checks for numbers with recurring digits
            if numerator % 11 == 0 or denominator % 11 == 0:
                continue

            # Checks for a recurring number between the two numbers
            if len(set(str(numerator) + str(denominator))) != 3:
                continue

            # Checks for non-trivial digit canceling fractions
            simplifed_frac = recurring_digits(numerator, denominator)
            if numerator / denominator == simplifed_frac:
                fractions.append((numerator, denominator))

    return fractions


def recurring_digits(num1, num2):
    str1 = str(num1)
    str2 = str(num2)

    if '0' in str1 or '0' in str2:
        return False

    if str1[0] == str2[0]:
        return int(str1[1]) / int(str2[1])
    elif str1[0] == str2[1]:
        return int(str1[1]) / int(str2[0])
    elif str1[1] == str2[0]:
        return int(str1[0]) / int(str2[1])
    elif str1[1] == str2[1]:
        return int(str1[0]) / int(str2[0])


if __name__ == '__main__':
    print(digit_canceling_fractions())
